// testcafe/lib/api/skip-js-errors/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareSkipJsErrorsOptions = exports.ensureSkipJsErrorsCallbackWrapped = exports.isSkipJsErrorsOptionsObject = exports.isSkipJsErrorsCallbackWithOptionsObject = void 0;
const make_reg_exp_1 = require("../../utils/make-reg-exp");
const client_function_builder_1 = __importDefault(require("../../client-functions/client-function-builder"));
const SKIP_JS_ERRORS_OBJECT_FUNCTION = `
                let { stack, pageUrl, message } = deps;

                return stack.test(err.stack) && pageUrl.test(err.pageUrl) && message.test(err.message);
        `;
function isSkipJsErrorsCallbackWithOptionsObject(obj) {
    return !!obj && typeof obj === 'object' && 'fn' in obj;
}
exports.isSkipJsErrorsCallbackWithOptionsObject = isSkipJsErrorsCallbackWithOptionsObject;
function isSkipJsErrorsOptionsObject(obj) {
    return !!obj && typeof obj === 'object' && !isSkipJsErrorsCallbackWithOptionsObject(obj);
}
exports.isSkipJsErrorsOptionsObject = isSkipJsErrorsOptionsObject;
function ensureSkipJsErrorsCallbackWrapped(options, dependencies = {}) {
    if (typeof options === 'function')
        return wrapSkipJsErrorsCallback(options, dependencies);
    return options;
}
exports.ensureSkipJsErrorsCallbackWrapped = ensureSkipJsErrorsCallbackWrapped;
function wrapSkipJsErrorsCallback(options, dependencies) {
    return { fn: options, dependencies };
}
function prepareSkipJsErrorsOptions(options) {
    options = ensureSkipJsErrorsCallbackWrapped(options);
    if (isSkipJsErrorsCallbackWithOptionsObject(options))
        return createSkipJsErrorsCallbackFunction(options);
    if (isSkipJsErrorsOptionsObject(options))
        return createSkipJsErrorsObjectFunction(prepareOptionsObject(options));
    return options;
}
exports.prepareSkipJsErrorsOptions = prepareSkipJsErrorsOptions;
function createSkipJsErrorsObjectFunction(deps) {
    deps.message = deps.message || new RegExp('');
    deps.stack = deps.stack || new RegExp('');
    deps.pageUrl = deps.pageUrl || new RegExp('');
    const func = new Function('err', SKIP_JS_ERRORS_OBJECT_FUNCTION);
    const callbackWrapper = wrapSkipJsErrorsCallback(func, { deps });
    return createSkipJsErrorsCallbackFunction(callbackWrapper);
}
function createSkipJsErrorsCallbackFunction({ fn, dependencies }) {
    const methodName = 'skipJsErrors handler';
    const options = { dependencies };
    return new client_function_builder_1.default(fn, options, {
        instantiation: methodName,
        execution: methodName,
    }).getCommand();
}
function prepareOptionsObject(opts) {
    opts.message = (0, make_reg_exp_1.parseRegExpString)(opts.message);
    opts.stack = (0, make_reg_exp_1.parseRegExpString)(opts.stack);
    opts.pageUrl = (0, make_reg_exp_1.parseRegExpString)(opts.pageUrl);
    return opts;
}
