// testcafe/lib/shared/errors/selector-error-ctor-callback.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCannotObtainInfoErrorCtor = exports.getNotFoundErrorCtor = exports.getInvisibleErrorCtor = void 0;
const Errors = __importStar(require("./index"));
function getInvisibleErrorCtor(elementName) {
    return !elementName ? 'ActionElementIsInvisibleError' : {
        name: 'ActionAdditionalElementIsInvisibleError',
        firstArg: elementName,
    };
}
exports.getInvisibleErrorCtor = getInvisibleErrorCtor;
function getNotFoundErrorCtor(elementName) {
    return !elementName ? 'ActionElementNotFoundError' : {
        name: 'ActionAdditionalElementNotFoundError',
        firstArg: elementName,
    };
}
exports.getNotFoundErrorCtor = getNotFoundErrorCtor;
function getCannotObtainInfoErrorCtor() {
    return 'CannotObtainInfoForElementSpecifiedBySelectorError';
}
exports.getCannotObtainInfoErrorCtor = getCannotObtainInfoErrorCtor;
function createErrorCtorCallback(errCtor) {
    // @ts-ignore
    const Error = typeof errCtor === 'string' ? Errors[errCtor] : Errors[errCtor.name];
    const firstArg = typeof errCtor === 'string' ? null : errCtor.firstArg;
    return (fn) => new Error(firstArg, fn);
}
exports.default = createErrorCtorCallback;
module.exports = exports.default;
